airbyte-ci/connectors/pipelines/pipelines/actions/environments.py

```diff
@@ -365,6 +365,25 @@ def with_python_connector_source(context: ConnectorContext) -> Container:
     return with_python_package(context, testing_environment, connector_source_path)
 
 
+async def apply_python_development_overrides(context: ConnectorContext, connector_container: Container) -> Container:
+    # Run the connector using the local cdk if flag is set
+    if context.use_local_cdk:
+        context.logger.info("Using local CDK")
+        # mount the local cdk
+        path_to_cdk = "airbyte-cdk/python/"
+        directory_to_mount = context.get_repo_dir(path_to_cdk)
+
+        context.logger.info(f"Mounting {directory_to_mount}")
+
+        # Install the airbyte-cdk package from the local directory
+        # We use --no-deps to avoid conflicts with the airbyte-cdk version required by the connector
```

Comment:
@clnoll great question! and you are exactly right it will ignore sub-dependency conflicts that may crop up.

e.g. if `airbyte-cdk` needs pandas v2.0.0 but `source-s3` need pandas v1.2.3 you wont see the error until you introduce the PR that explicitly bumps `source-s3`'s `airbytec-dk` version.

The alternative is that we do not include `--no-deps` and to test local cdk changes a cdk developer would have to
1. bump the library version number in the pyproject.toml for `airbyte-cdk`
2. Ensure/bump the `airbyte-cdk` library version in the pyproject.toml for a connector you would like to test with to match
3. Run `airbyte-ci --use-local-cdk --name=source-s3 test`

The reason I did not choose this route was because
1. Its an extra step / commit to test different connectors (you have to update their dependency file)
2. I **assumed** that the PR that released a new cdk version would not be the same PR that bumped a connector to use the new version. That assumption let me believe the dependency conflicts would be found later in the second PR since you would be running test without `--use-local-cdk`.

@clnoll more than happy to change the approach, what are your thoughts?

Reply:
:+1: thanks! I want to take a closer look at the code and understand why we’re stuck with this tradeoff between --no-deps and extra steps, given that the original implementation of the local cdk flag didn’t have that tradeoff. This isn’t a super common scenario but it’s common enough that it doesn’t feel quite right to me to skip the deps.

I hear you re the fact that conflicts would be caught before release, but I think that rooting out the issue early is important. Correct me if I'm wrong, but with the current implementation, to fix a dependency conflict I believe we’d have to toggle back and forth between updating the CDK and updating the connector until we settle on a version that agrees with both. Some conflicts could be resolved by doing that locally, but sometimes there are issues that will only crop up in certain development environments. i.e. you’ll see them in CI but not during local development, especially if it involves something like numpy or pandas that has a C dependency - and conflicts related to those libraries are the biggest sources of these problems that I've seen.

Reply:
> I want to take a closer look at the code and understand why we’re stuck with this tradeoff between --no-deps and extra steps
Ah whoops forgot to put a link to part of the why:
https://github.com/jazzband/pip-tools/issues/215

> the original implementation of the local cdk flag didn’t have that tradeoff
I can answer that!

So the original implementation used `sed` to edit the setup.py file using find and replace.

This approach is both brittle and a victim of the python ecosystem where there are approximately 4 "standard" ways in 2023 to define your dependencies
1. Using requirements.txt
2. Using setup.py
3. Using pyproject.toml
4. Using pyproject.toml + poetry.lock

Were currently planning on forcing python connectors to all move to the "pyproject.toml + poetry.lock" method but thats still loose plans and today were stuck with 4 possible ways that the `airbyte-cdk` dependency is declared.

To continue to use the `sed` approach we would have to have parsing and local override code for all 4 of these files.

However, because they all end up using `pip` at the end of the day we chose to 
1. Let the connector install using any of the above methods (for now)
2. Call `pip install path/to/airbyte-cdk/python` after the fact.

As the trade off is to either
1. Ignore connector vs airbyte-cdk dependencies conflict during cdk testing
2. Or, require you to update a connectors dependency file when testing the cdk

> Correct me if I'm wrong, but with the current implementation, to fix a dependency conflict I believe we’d have to toggle back and forth between updating the CDK and updating the connector until we settle on a version that agrees with both

🤔 I dont think you'd have to toggle back and forth.

Let me explain

Lets assume when your updating a connector to use a new version of the cdk you have this goal in mind
* **Upgrade to the latest possible version of cdk that wont require a massive refactor**

WIth this you'll likely target latest cdk.
* If theres a dependency conflict:
    * you will look at upgrading your own depenencies first
* If you cant upgrade your dependencies (You need an old version of pandas or something):
    * you will likely walk back a version of the airbyte-cdk and try again.
    
I dont think a connector dev would be messing with cdk dependency versions
and vice versa, I dont think a cdk dev would be messing with a connector version (if theres a major conflict, try testing with another connector)

@clnoll Is that a fair way to think about it?


> but sometimes there are issues that will only crop up in certain development environments. i.e. you’ll see them in CI but not during local development, especially if it involves something like numpy or pandas that has a C dependency - and conflicts related to those libraries are the biggest sources of these problems that I've seen.

Were hoping this method solves this very issue where `airbyte-ci connectors build`, `airbyte-ci connectors test`, and `airbyte-ci connectors publish` both use the same, repeatable (dockerized) environment both on your machine and in CI

Reply:
Thanks so much for the details, @bnchrch!

I agree with your description of how a connector dev would be thinking about this situation. Re the CDK dev perspective - we do test changes on existing connectors and I could see us wanting to make sure that a package upgrade was appropriate for specific connectors, particularly if it's a popular connector, or if a CDK update was made with solving problems specific to those connectors.

But, if we're confident that `airbyte-ci` is reproducing the prod environment I think we're good to just deal with this scenario locally.